### ISSUE TITLE
SQL syntax error when using concat and parentheses in backend list

### ISSUE THREAD

**Issue by leoquijano** [octobercms/october#2449]
Hi,

I'm trying to use the following column definitions for a backend model in a PGSQL site:

columns:
    id:
        label: ID
        searchable: true
    amount_due:
        label: Amount Due
    currency_code:
        label: Currency
    description:
        label: Description
    payments:
        label: Payments
        relation: payments
        select: 'concat("#", id, " - ", amount, "(", currency_code, ")")'
    ...

Please note the use case, I'd like to display the currency_code using parentheses. Like this: 500 (USD). However, the SQL parser is not getting it correctly. This is what the above select generates:

```
"#" || id || " - " || amount || "(" || currency_code || "")
```

I noticed that this in an old issue, as referenced in #883. However, testing the above string in the parseConcat function, directly, is removing my last closing parentheses:

``` php
public function parseConcat($sql)
{
    return preg_replace_callback('/(?:group_)?concat\(([^)]+)\)(?R)?/i', function($matches){
        if (!isset($matches[1])) {
            return $matches[0];
        }

        // This is a group_concat() so ignore it
        if (strpos($matches[0], 'group_') === 0) {
            return $matches[0];
        }

        $concatFields = array_map('trim', explode(',', $matches[1]));

        switch ($this->driver) {
            default:
            case 'mysql':
                return $matches[0];

            case 'pgsql':
            case 'postgis':
            case 'sqlite':
                return implode(' || ', $concatFields);
        }
    }, $sql);
}
```

I tested the above function using phptester.net, and yes, it is eating the last closing ")".

**Comment by Patroklo** [comment]
Why you don't make in this case a "false" attribute called whatever and return that data from the php model?

Like:

```php
    public function getWhateverAttribute()
    {
        return '('.$this->getAttribute('id').') - '.$this->getAttribute('amount')....;
    }
```

It should work for lists

**Comment by leoquijano** [comment]
It is a good workaround. I actually replaced CONCAT with ||.

But I think the problem is more general: the parser is eating the last ")". It's not concat-specific.

**Comment by Patroklo** [comment]
A question about your defined concat... I have just seen that you are using an string instead of a function there. Why is that?

I mean, you don't have to use an string there, but values that will be parsed into a select statement, if you use a string you are making october do something like this: `select 'concat( cacafuti ) from tablefuti` and will throw an exception. 

What it seems to me that is happening here is that October it's getting the mysql error (for instance we will use the sql `select 'concat(title,' ',comment)' from table_derp` to explain this case), launching this query will throw an error like `blah blah near '')' from table_derp' at line 1`. And I think that october its getting this as a valid query value instead of an exception and it throws another exception because it can't handle it apropriatly? I'll have to check deeper in the code to be sure about that, but it seems plausible...

tl;dr: [why concat string?! Manual says not to use it like that!](http://octobercms.com/docs/backend/lists#column-select)

**Comment by leoquijano** [comment]
1. I'm using PostgreSQL, not MySQL

2. I did run a test on the regular expression (```parseConcat``` function) and it doesn't pick it right. Try it yourself: 

```
/(?:group_)?concat\(([^)]+)\)(?R)?/i
```

Either October CMS supports SQL or it doesn't. Using it as a function is valid SQL, so why wouldn't October be able to handle it?

**Comment by Patroklo** [comment]
Problem it's that your example is working for me (at least my version, probably I have something different). What is exactly your error?

**Comment by leoquijano** [comment]
Take a look:

https://regex101.com/r/gGbWZi/1

**Comment by Patroklo** [comment]
All right, now I see the problem and why it works only for MySQL 8D It's using the concat without regex and the non mysql drivers are using the other one.

**Comment by Patroklo** [comment]
Le fu~ this has a lot of inside complications and I'm not a skilled regex developer. Right now I have this that somewhat (aka, have tried only two examples) works `((?:group_)?concat)?(\(((?:[^\(\)]++|(?R))*)\))`

**Comment by leoquijano** [comment]
It would probably be better to use an actual SQL parser, as regexps have a limit when strings get more complex.

**Comment by Patroklo** [comment]
Well... I know this sql parser, but it seems overkill here... https://github.com/greenlion/PHP-SQL-Parser

**Comment by leoquijano** [comment]
Yeah, and even then, it's only for MySQL.

Why was October doing that ```parseConcat```, anyway?

**Comment by daftspunk** [comment]
This will be fixed in October Build 381+
